causync.py

```diff
@@ -0,0 +1,374 @@
+            return dirdate
+        except ValueError as e:
+            self.logger.error(e)
+            exit(1)
+
+    def find_latest_backups(self, dirnames, count=5):
+        """ Returns the latest daily backup directory names. """
+
+        if len(dirnames) == 0:
+            return []
+
+        # extract dates from directory names and sort them
+        dirdates = [ self.get_dirdate(d) for d in dirnames ]
+        dirdates.sort(reverse=True)
+
+        # determine list length (if len < delete count, we don't do anything)
+        list_len = count if len(dirdates) >= count else len(dirdates)
+        # get a reverse sorted list of dirdates (descending by date)
+        latest_dates = dirdates[:list_len]
+        # convert them to string for rsync --link-dest (example: datetime obj becomes 'YYMMHH' string)
+        latest_dates = [ i.strftime(config.DATE_FORMAT) for i in latest_dates ]
+        # join each one with the destination directory (example: '/path/dest/sourcedir_YYMMHH'
+        latest_dates = [ os.path.join(self.dst_abs, i) for i in latest_dates ]
+
+        return latest_dates
+
+    def find_old_backups(self, dirnames, ival='daily', count=5):
+        """ Returns old backups we should delete.
+            The time interval is specified by 'ival'. Examples: daily, weekly, monthly, yearly.
+        """
+
+        curdate = datetime.now()
+        multiplier = timedelta(days=self.config.BACKUP_MULTIPLIERS[ival])
+
+        # extract dates from directory names and sort them
+        dirdates = list(sorted([ self.get_dirdate(d) for d in dirnames ]))
+
+        (keep, delete) = (list(), list())
+
+        for d in dirdates:
+            m = multiplier * count + multiplier
+            keepdate = curdate - m
+
+            if ival == 'yearly' and d.day != 1 and d.month != 1:
+                continue
+            elif ival == 'monthly' and d.day != 1:
+                continue
+            elif ival == 'weekly' and d.weekday() != 0:
+                continue
+
+            if d > keepdate:
+                keep.append(d)
+            else:
+                delete.append(d)
+
+        delete.sort(reverse=True)
+        dirdates.sort(reverse=True)
+        keep.sort(reverse=True)
+
+        return keep, delete
+
+    def run_cleanup(self):
+        """ Deletes old backups.
+            You can set how many you want to keep for each date/time interval in config.py.
+            This function is executed when the task argument is 'cleanup'.
+        """
+
+        listdir = os.listdir(self.dst_abs)
+
+        yearly = self.find_old_backups(listdir, 'yearly', self.config.BACKUPS_TO_KEEP['yearly'])
+        monthly = self.find_old_backups(listdir, 'monthly', self.config.BACKUPS_TO_KEEP['monthly'])
+        weekly = self.find_old_backups(listdir, 'weekly', self.config.BACKUPS_TO_KEEP['weekly'])
+        daily = self.find_old_backups(listdir, 'daily', self.config.BACKUPS_TO_KEEP['daily'])
+
```

Comment:
This looks quite chaotic, you simply need the aggregated list of delete: sorted(set(yearly[1] + monthly[1] + weekly[1] + daily[1])). Then you can delete them all...